crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2321,23 +2315,30 @@ impl<'db> Type<'db> {
                 //     def __get__(self, instance: object, owner: type | None = None, /) -> MethodType: ...
                 // ```
 
-                let first_argument_is_none =
-                    arguments.first_argument().is_some_and(|ty| ty.is_none(db));
-
-                let signature = Signature::new(
-                    Parameters::new([
-                        Parameter::new(
-                            Some("instance".into()),
-                            Some(Type::object(db)),
-                            ParameterKind::PositionalOnly { default_ty: None },
-                        ),
-                        if first_argument_is_none {
+                let not_none = Type::none(db).negate(db);
+                let overloads = Overloads::from_overloads([
+                    Signature::new(
+                        Parameters::new([
+                            Parameter::new(
+                                Some("instance".into()),
+                                Some(Type::none(db)),
+                                ParameterKind::PositionalOnly { default_ty: None },
+                            ),
```

Comment:
All of these updates will help separate call binding into two phases, since all of the `Signature`s that we construct no longer depend on the inferred argument types.  (We still have special case logic to update the return type of the call based on argument types, but that won't conflict with the two-phase change)